feat: set cjson.array_mt on decoded JSON arrays

thibaultcha opened the pull request:
> Built on top of #26, but opened against master.

This implements a new config value to ensure that decoded JSON Arrays will be subsequently re-encoded by this module as JSON Arrays. The behavior can be configured via a new configuration flag:
```lua
cjson.decode_empty_array_with_array_mt("on"|"off")
```

Example usage:
```lua
local cjson = require "cjson"

local my_json = [[{"my_array":[]}]]
local t = cjson.decode(my_json)
cjson.encode(t) -- {"my_array":[]}

-- now, if this behavior is disabled
cjson.decode_empty_array_with_array_mt("off")

local my_json = [[{"my_array":[]}]]
local t = cjson.decode(my_json)
cjson.encode(t) -- {"my_array":{}}
```

Fixes #25 
Fixes openresty/openresty#193

spacewander commented:
@agentzh 
Do you have spare time to review this pr? We are in need of this feature too.

agentzh commented:
@spacewander Yeah, will have a look.

Also, it would be great if we can measure if there's any performance difference when decoding large and complicated JSON data samples.

agentzh commented on lua_cjson.c:
```diff
@@ -328,6 +331,16 @@ static int json_cfg_encode_empty_table_as_object(lua_State *l)
 }
 
+/* Configures how to decode arrays */
+static int json_cgf_decode_array_with_array_mt(lua_State *l)
```
Typo? "json_cfg_xxx"?

agentzh commented on lua_cjson.c:
```diff
@@ -713,22 +727,39 @@ static void json_append_data(lua_State *l, json_config_t *cfg,
-                json_append_object(l, cfg, current_depth, json);
+                if (has_metatable) {
+                    lua_getmetatable(l, -1);
+                    lua_pushlightuserdata(l, &json_empty_array);
```
Are you sure it's "json_empty_array" instead of "json_array"?

thibaultcha replied:
Yes, probably. This code was written on top of `master`, and later rebased on top of #26, which ontroduces `json_array`.

thibaultcha replied:
Had another look at this - this code is correct, it is to preserve backwards compatibility with `cjson.empty_array_mt`.

thibaultcha commented on lua_cjson.c:
```diff
@@ -713,22 +727,39 @@ static void json_append_data(lua_State *l, json_config_t *cfg,
+                if (has_metatable) {
+                    lua_getmetatable(l, -1);
+                    lua_pushlightuserdata(l, &json_empty_array);
+                    lua_rawget(l, LUA_REGISTRYINDEX);
```
Yes, that is fine with me - although outside of the scope of this PR as this part of the code predates this PR (but I'll include it nonetheless to reduce branching) - Thanks!

thibaultcha commented:
@spacewander @agentzh I just pushed @spacewander's suggestion to #26 and fixed the typo you spotted @agentzh in this PR. As before, this PR is rebased on top of #26 (it has both the `cjson.array_mt` and the `cjson.decode()` commits). Glad this is getting interest, if I can do anything to push it forward, let me know!

@agentzh I will provide such samples as soon as I can manage

thibaultcha commented:
@agentzh From my preliminary tests, this change does not seem to have any measurable impact on performance. By running 5 `decode` operations on a large JSON document prior and after this patch, I get the following on a MacBook Pro 2016 2.7GHz i7):

```
cjson.decode (old) avg: 11.754357s
cjson.decode (new) avg: 11.416297s
```

Here is my benchmarking script:

1. `$ cd lua-cjson`
2. `$ git checkout [this branch]`
3. `$ make` (also need to have a `cjson.so` without thi schange in `LUA_PACKAGE_CPATH`)
4. `$ wget https://raw.githubusercontent.com/zemirco/sf-city-lots-json/master/citylots.json`
5. `$ luajit ./bench.lua`

Here is bench.lua:
```lua
local socket = require "socket"

local f = assert(io.open("./citylots.json", "r"))
local str = assert(f:read("a"))

local N = 5

do
  local cjson = require "cjson"

  local avg = 0
  local tstart = socket:gettime()
  for i = 1, N do
    cjson.decode(str)
    avg = avg + socket:gettime() - tstart
  end

  print(string.format("cjson.decode (old) avg: %fs", avg / N))
end

package.loaded["cjson"] = nil
collectgarbage()
collectgarbage()

do
  package.cpath = "./?.so;" .. package.cpath
  local cjson = require "cjson"

  local avg = 0
  local tstart = socket:gettime()
  for i = 1, N do
    cjson.decode(str)
    avg = avg + socket:gettime() - tstart
  end

  print(string.format("cjson.decode (new) avg: %fs", avg / N))
end
```

Let me know your thoughts!

agentzh commented:
@thibaultcha I think you should really move `avg = avg + socket:gettime() - tstart` out of the loop? Otherwise you are counting `socket:gettime()`'s performance inside the loop as well.

thibaultcha commented:
@agentzh Indeed. Well, doing so yields the same difference between the old and the new implementation:
```
cjson.decode (old) avg: 4.071810s
cjson.decode (new) avg: 4.004254s
```

thibaultcha commented:
(I think the biggest issue was that `tstart` was not inside the loop, while `tend` was calculated from there - effectively not giving an "average"... The order of magnitude in the performance impact is still the same though)

thibaultcha commented:
Two more runs with `N = 20` gave the following results:
```
cjson.decode (old) avg: 4.168498s
cjson.decode (new) avg: 4.196873s
```

```
cjson.decode (old) avg: 4.151831s
cjson.decode (new) avg: 4.151734s
```

thibaultcha commented:
(I have updated the script to reflect the state of this benchmark script as it ran the latest results I posted above)

agentzh commented:
@thibaultcha That looks great! Thanks for your hard work on this.

thibaultcha commented:
@agentzh Appreciated! Let me know if there is anything I can assist here with. Maybe we should merge #26 and this PR separately, to keep the history atomic...

agentzh commented on README.md:
```diff
@@ -122,3 +157,38 @@ encode_number_precision
+--------------------------------
+**syntax:** `cjson.decode_array_with_array_mt(enabled)`
+
+**default:** true
```
Should we turn this off by default for the best backward compatibility?

thibaultcha replied:
I am ok with this, but I think the default behavior really is broken, and that backwards compatibility shouldn't suffer if we enable this by default. I'll turn it off anyways just to be safe.

agentzh commented:
@thibaultcha Will you please rebase to the latest master? Thanks!

agentzh commented on README.md:
```diff
@@ -122,3 +157,38 @@ encode_number_precision
+If disabled, JSON Arrays will be decoded to plain Lua tables, without
+the `array_mt` metatable.
+
+The `enabled` argument can be a boolean, or a string (`on` or `off`).
```
I think we should only support boolean values here, at least in the documentation.

thibaultcha replied:
We refer to the `"on"` and `"off"` variants of the arguments in the documentation for [encode_empty_table_as_object()](https://github.com/openresty/lua-cjson#encode_empty_table_as_object) already. Should we remove such mentions from there as well?

agentzh replied:
@thibaultcha Yes, please.

agentzh commented on README.md:
```diff
@@ -122,3 +157,38 @@ encode_number_precision
 [Back to TOC](#table-of-contents)
+
+decode_array_with_array_mt
+--------------------------------
```
Better remove the excessive dashes here.

thibaultcha replied:
Oops, definitely an oversight, thanks for the catch!

thibaultcha commented:
@agentzh I have rebased on top of master, updated the commit message, and addressed your concerns (disable by default for backwards compatibility concerns)!

agentzh commented:
@thibaultcha Merged. Thanks!
